wgpu-core/src/device/global.rs

```diff
@@ -2280,32 +2173,33 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
     ///
     /// Return `all_queue_empty` indicating whether there are more queue
     /// submissions still in flight.
-    fn poll_devices<A: HalApi>(
+    fn poll_device<A: HalApi>(
         &self,
         force_wait: bool,
         closures: &mut UserClosures,
     ) -> Result<bool, WaitIdleError> {
-        profiling::scope!("poll_devices");
+        profiling::scope!("poll_device");
 
         let hub = A::hub(self);
         let mut devices_to_drop = vec![];
         let mut all_queue_empty = true;
         {
-            let mut token = Token::root();
-            let (device_guard, mut token) = hub.devices.read(&mut token);
+            let device_guard = hub.devices.read();
 
             for (id, device) in device_guard.iter(A::VARIANT) {
                 let maintain = if force_wait {
                     wgt::Maintain::Wait
                 } else {
                     wgt::Maintain::Poll
                 };
-                let (cbs, queue_empty) = device.maintain(hub, maintain, &mut token)?;
+                let fence = device.fence.read();
+                let fence = fence.as_ref().unwrap();
+                let (cbs, queue_empty) = device.maintain(hub, fence, maintain)?;
                 all_queue_empty = all_queue_empty && queue_empty;
 
-                // If the device's own `RefCount` clone is the only one left, and
+                // If the device's own `RefCount` is the only one left, and
                 // its submission queue is empty, then it can be freed.
-                if queue_empty && device.ref_count.load() == 1 {
+                if queue_empty && device.is_unique() {
                     devices_to_drop.push(id);
                 }
```

Comment:
This doesn't seem right. If the sole `Arc<Device>` remaining is the device's entry in `Hub::devices`, it should not be freed yet - it can still be referred to by that id.